src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+        self.image = image
+
+    @classmethod
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def install_deps(cls) -> None:
+        """Install dependencies."""
+
+        logger.info("Installing charm dependencies.")
+
+        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
+        execute_command(["/usr/bin/apt", "remove", "-qy", "lxd", "lxd-client"], check=False)
+        execute_command(["/usr/bin/snap", "install", "lxd", "--channel=latest/stable"])
+        execute_command(["/usr/bin/snap", "refresh", "lxd", "--channel=latest/stable"])
+        execute_command(["/snap/bin/lxd", "waitready"])
+        execute_command(["/snap/bin/lxd", "init", "--auto"])
+        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
+        execute_command(["/snap/bin/lxc", "network", "set", "lxdbr0", "ipv6.address", "none"])
+        execute_command(
+            [
+                "/usr/bin/apt",
+                "install",
+                "-qy",
+                "cpu-checker",
+                "libvirt-clients",
+                "libvirt-daemon-driver-qemu",
+            ],
+        )
+        cls.runner_bin_path.parent.mkdir(parents=True, exist_ok=True)
+
+        logger.info("Finished installing charm dependencies.")
+
+    def get_latest_runner_bin_url(self, os_name: str = "linux", arch_name: str = "x64") -> str:
+        """Get the URL for the latest runner binary.
+
+        The runner binary URL changes when a new version is available.
+
+        Returns:
+            The URL to the latest runner binary.
+        """
+
+        runner_bins: RunnerApplicationList = []
+        if isinstance(self.path, GitHubRepo):
+            runner_bins = self._github.actions.list_runner_applications_for_repo(
+                owner=self.path.owner, repo=self.path.repo
+            )
+        elif isinstance(self.path, GitHubOrg):
+            runner_bins = self._github.actions.list_runner_application_for_org(org=self.path.org)
+
+        logger.debug("Response of runner binary list: %s", runner_bins)
+
+        for bin in runner_bins:
+            if bin.os == os_name and bin.architecture == arch_name:
+                return bin.download_url
+
+        raise RunnerBinaryError(
+            f"Unable to download GitHub runner binary for {os_name} {arch_name}"
+        )
+
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def update_runner_bin(self, download_url: str) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Args:
+            download_url (str): URL to download the runner binary.
+
+        TODO:
+            Deletion first cause no runner binary on download failure. See if this cause problems.
+        """
+        logger.info("Downloading runner binary from: %s", download_url)
+
+        # Remove any existing runner bin file
+        if self.runner_bin_path.exists():
+            self.runner_bin_path.unlink()
```

Comment:
Should the unlock happen just after the download? Also, would it be worth creating the file for the new binary first, then unlicking and then rename the file? That way, the time no binary file for the runner is available is minimised. Not sure if it matters, e.g., what would happen if the file is temporarily unavailable

Reply:
This was my initial thought, but then I realized if there is a updated version of the binary, and the charm fails to download it. Should the charm use the old version? or wait until it can download the newest version? 
I am unsure as this might have security implications (e.g., security bug fixes in new version).

Reply:
Yes that is a good point, that was one of the security feedback items to harden the update process. Maybe it would be better to delete the binary and stop accepting runs as soon as a new binary is released? And log/ alert it on download failure until manual intervention?

Reply:
I'm guessing it would refuse the run actions at that point, the theory is that there may be a security vulnerability with the runner binary we have given there is a new one, so the trade-off here is security versus usability. E.g., the more secure way is to stop using the current binary as soon as we become aware of a new binary. The usability perspective is to keep using a potentially vulnerable binary until we can successfully download the new binary

Reply:
As long as we accept that in the exceptional situation where there is a security vuln AND the new version is non downloadable (either due to an incident on our side or due to an issue on the site hosting the new version), the runners aren't available
Then it's fine